flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannel.java

```diff
@@ -338,10 +338,13 @@ void releaseAllResources() throws IOException {
 
     @Override
     void announceBufferSize(int newBufferSize) {
-        checkState(!isReleased, "Channel released.");
-
-        ResultSubpartitionView subpartitionView = checkNotNull(this.subpartitionView);
-        subpartitionView.notifyNewBufferSize(newBufferSize);
+        ResultSubpartitionView view = this.subpartitionView;
+        // if releaseAllResources would be called from the mailbox thread it is possible that
```

Comment:
I am not entirely sure if the description here is correct, or better said if it is precise.

I believe calling any of the methods from a mailbox thread is not the reason for the problem. Correct me if I am wrong, but I believe the problem is exactly the same as described in `LocalInputChannel:198-204`.

Reply:
BTW, don't we have a bigger problem, that we try to perform debloating even if a `SingleInputGate` might've finished already? What happens if we have two input gates and one of the gates finishes early (receives the end of partition), while the other processes records continuously? Could you try that out?

Reply:
The key problem here is following:

- The mailbox takes EndOfPartitionEvent for handling
- At the same time, timerService submit buffer debloater to the mailbox
- The mailbox release all resources for the channel because of EndOfPartitionEvent
- the mailbox submit suspend mailbox task to mailbox because all channels  were ended
- The mailbox takes the next task which is buffer debloater and we have the problem describing in this task

Reply:
Maybe I don't fully understand the problem about finished `SingleInputGate`, but as I understand after this ticket we will just send nothing to already finished channels. So it should work totally fine. 
Or is your question about how correctly can we calculate buffer size if some of the gates would be finished? ( I actually think it should also work especially after your ticket about multiply gates)

Reply:
Yes, exactly the problem you describe in the steps above is what I meant. It has very little to do with the mailbox. It would be the same if the steps above were called outside of the mailbox. It is a problem of closing the channel on `EndOfPartition`.

Per finished input gates. Yes, we will ignore it for local channels. What about remote channels? What happens if you try to send an announcement on a closed channel? Moreover, why should we bother with gates that have finished? We should not perform any operations for them.

Reply:
I updated PR one more time. But my current solution considers that it is impossible to call `announceBufferSize` and `releaseAllResources` in parallel. But I actually think that it is possible. Right now, I am trying to figure it out but if you know the answer, please, share it with me.

Reply:
> What happens if you try to send an announcement on a closed channel?
As I see, it is not a problem for the remote channel because we remove the listener on releasing, and when we send something to the closed channel we just ignore it.(at least I see it in my brief test)

Reply:
> But my current solution considers that it is impossible to call announceBufferSize and releaseAllResources in parallel

Please confirm that this may happen only when cancelling or failing a task externally, because we `closeAllInputGates` from the `TaskCanceler` thread. I've looked into it and I think that:
1. It's the same/similar for all of the methods. The cancellation may happen at any given time, so taking e.g. `RemoteInputChannel#resumeConsumption` as an example, the channel might be released after the `checkState` condition.
2. It is rather unlikely to cause problems. It very much depends on timing.
3. Even if we fail (either on the `checkState` or while sending messages to the channel (I believe it might happen. I don't know which _listener_ you had in mind) it should not matter much. We are failing anyway and this exception should be ignored. We are already in FAILED/CANCELLING state (see `Task:765-821`).

Please let me know what you think.

Reply:
> Please confirm that this may happen only when cancelling or failing a task externally,

I confirm that in my opinion, it is the only place where it can happen.

In general, I agree that it is the same situation as with other similar methods (like `RemoteInputChannel#resumeConsumption`) and I agree that it should not matter much because it means that job is  already failing. So perhaps, my last changes should work fine.